### PR TITLE
d2l-list - Selecting items render - (2/5)

### PR DESCRIPTION
**Pull request by m6jones**
#  [&larr; PREV ](https://github.com/BrightspaceUI/core/pull/180) : PR NAV : [NEXT &rarr;](https://github.com/BrightspaceUI/core/pull/211)

# About this PR

This PR adds the style of the selection. So everything related to rending what the selecting list items is part of this PR.

## Out of scope:

- Readme update
- links
- Draggable
- Multiple Columns
- Anyway to use the information that an item was selected

# Useful Links

## Demo: 

This demo is a demo of the PR series:
http://i-00000089.openstack.d2l:8081/components/@brightspace-ui/core/components/list/demo/list.html

## Readme: 

https://github.com/BrightspaceUI/core/blob/majones/d2lListReadme/components/list/README.md

## Design: 

http://design.d2l/components/lists/

### CODE REVIEW

File: components/inputs/checkbox-shared-styles.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+		border-width: 1px;
+	}
+	input[type="checkbox"]:hover,
+	input[type="checkbox"]:focus,
+	input[type="checkbox"].d2l-input-checkbox-focus {
+		border-color: #006fbf;
+		border-width: 2px;
+		outline-width: 0;
+	}
+	`;
```
Comment (dlockhart) [inline components/inputs/checkbox-shared-styles.js]: I'm hoping to pull all the input styles into core sometime next week. Not sure what your timeline is for this stuff, but either way, there'll be checkbox stuff soon.

Comment (m6jones) [reply components/inputs/checkbox-shared-styles.js]: Awesome!! I was going to ask how this should be organized but having the shared style is good for this component since I need to modify how the label position.

File: components/inputs/checkbox-shared-styles.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+	}
+	input[type="checkbox"]:checked {
+		background-image: url("data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A//www.w3.org/2000/svg%22%20width%3D%2224%22%20height%3D%2224%22%20viewBox%3D%220%200%2024%2024%22%3E%3Cpath%20fill%3D%22%23565A5C%22%20d%3D%22M8.4%2016.6c.6.6%201.5.6%202.1%200l8-8c.6-.6.6-1.5%200-2.1-.6-.6-1.5-.6-2.1%200l-6.9%207-1.9-1.9c-.6-.6-1.5-.6-2.1%200-.6.6-.6%201.5%200%202.1l2.9%202.9z%22/%3E%3C/svg%3E%0A");
+	}
+	input[type="checkbox"]:indeterminate {
+		background-image: url("data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A//www.w3.org/2000/svg%22%20width%3D%2224%22%20height%3D%2224%22%20viewBox%3D%220%200%2024%2024%22%3E%3Cpath%20fill%3D%22%23565A5C%22%20d%3D%22M7.5%2C11h9c0.8%2C0%2C1.5%2C0.7%2C1.5%2C1.5l0%2C0c0%2C0.8-0.7%2C1.5-1.5%2C1.5h-9C6.7%2C14%2C6%2C13.3%2C6%2C12.5l0%2C0%0A%09C6%2C11.7%2C6.7%2C11%2C7.5%2C11z%22/%3E%3C/svg%3E%0A");
+	}
+	input[type="checkbox"],
+	input[type="checkbox"]:hover:disabled {
+		background-color: #f9fafb;
```
Comment (dbatiste) [inline components/inputs/checkbox-shared-styles.js]: These colors don't match out existing checkbox colors, and they aren't part of our palette either.  I am wondering if we can be consistent?  ref: [inputs](https://github.com/BrightspaceUI/inputs/blob/master/d2l-input-checkbox.js#L85)

File: components/inputs/checkbox-shared-styles.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+		border-color: #d3d9e3;
+		border-width: 1px;
+	}
+	input[type="checkbox"]:hover,
+	input[type="checkbox"]:focus,
+	input[type="checkbox"].d2l-input-checkbox-focus {
+		border-color: #006fbf;
```
Comment (dbatiste) [inline components/inputs/checkbox-shared-styles.js]: This appears to be `d2l-color-celestine` - but I think we should check if this is consistent.

File: components/list/list.js [hunk 4]
```diff
@@ -4,7 +4,9 @@ class List extends LitElement {
 	static get properties() {
 		return {
 			dividerMode: { type: String, attribute: 'divider-mode' },
-			dividerExtend: {type: Boolean, attribute: 'divider-extend'}
+			dividerExtend: { type: Boolean, attribute: 'divider-extend'},
+			hoverEffect: {type: Boolean, attribute: 'hover-effect'},
+			selectable: {type: Boolean }
```
Comment (dbatiste) [inline components/list/list.js]: I think we are making an assumption (with `selectable:Boolean`) that selection is always multiple/checkboxes.  The MVC list supports single selection, which I think was used very rarely.  This _might_ be ok, but it's worth calling out, given that it has consequence for the public API and changing later could be painful.

File: components/list/list-item.js [hunk 5]
```diff
@@ -26,65 +29,104 @@ class ListItem extends RtlMixin(LitElement) {
...
 				align-self: flex-start;
-				display: flex;
+				display: grid;
 				flex-grow: 0;
+				grid-auto-columns: 1fr;
+				grid-auto-flow: column;
+				grid-gap: 0.3rem;
+				margin: 0.15rem 0;
+				position: relative;
+				z-index: 150;
```
Comment (dbatiste) [inline components/list/list-item.js]: How was `150` determined as the right z-index value?

Changing positioning and `z-index` of the `::slotted` items worries me.  When possible, I think I'd rather do as much locally as possible and not apply styles to user-land DOM elements.

File: components/list/list-item.js [hunk 6]
```diff
@@ -26,65 +29,104 @@ class ListItem extends RtlMixin(LitElement) {
...
 				margin-right: 0;
 			}
-			:host([illustration-outside]) ::slotted([slot="illustration"]) {
-				margin-bottom: 18px;
-				margin-top: 18px;
-			}
 			::slotted([slot="actions"]) {
 				align-self: flex-start;
-				display: flex;
+				display: grid;
```
Comment (dbatiste) [inline components/list/list-item.js]: What is the rationale for switching to grid - is it to achieve vertical centering of the input?  Are there issues in IE11 or Edge?

Comment (m6jones) [reply components/list/list-item.js]: It makes it better to add multiple actions that have the correct responsive behaviour. I'll need to test it in edge.

It shouldn't be vertically centering. So that is a mistake I'll have to fix.

It also makes spacing easier to define with the grid-gap.

File: components/list/list-item.js [hunk 7]
```diff
@@ -137,11 +180,17 @@ class ListItem extends RtlMixin(LitElement) {
 	}
 
 	render() {
-		const illustrationSlot = html`<slot name="illustration"></slot>`;
+		const contentId = getUniqueId();
+		const checkBoxId = getUniqueId();
+		const illustrationSlot = html`
+			<input @change="${this._handleChange}" class="d2l-list-item-checkbox" type="checkbox" id="${checkBoxId}">
```
Comment (dbatiste) [inline components/list/list-item.js]: Perhaps we should only render the `input` if the item is selectable.  However that would require the items "know" whether they are selectable or not... and that would require either the list-item reaching out to check this on it's parent list component (not a huge fan of this), or adding a property to list-item.

File: components/inputs/checkbox-shared-styles.js (hunk 2 above)
Comment (m6jones) [reply components/inputs/checkbox-shared-styles.js]: I copied and pasted the style. Not sure how this happened. But yeah, I want it to be the same as the current inputs. So 100% needs to be consistent.

File: components/list/list.js (hunk 4 above)
Comment (m6jones) [reply components/list/list.js]: This is interesting. I did make the assumption that if the list has selectable element that all elements are selectable. I'll have to talk to Rebecca and Colin about it.

I did make radio selection not part my work but that will be added in the future. Do you think I should make it: `selectable: [checkbox|radio|none(default)]`

File: components/list/list-item.js (hunk 7 above)
Comment (dbatiste) [inline components/list/list-item.js]: I think this should be binding to the `checked` property of the list-item. I suspect we'll want `.checked=${this.checked}`.

Comment (m6jones) [reply components/list/list-item.js]: Yes, I considered both of those as well. All of them seem like they will have a performance hit. So I decided to go with the cleanest code approach for now.

File: components/list/list-item.js [hunk 8]
```diff
@@ -16,6 +18,7 @@ class ListItem extends RtlMixin(LitElement) {
 	static get properties() {
 		return {
 			breakpoints: { type: Array },
+			checked: { type: Boolean, reflect: true },
```
Comment (dbatiste) [inline components/list/list-item.js]: Not sure whether this should be `checked` or `selected`.  The `checked` term is specific to checkboxes, and if there were a such thing as single selection, it wouldn't be quite right I think.

File: components/list/list.js [hunk 9]
```diff
@@ -17,25 +19,49 @@ class List extends LitElement {
...
-				--d2l-list-item-content-padding: 0 18px;
+				--d2l-list-item-content-padding-side: 0.9rem;
+			}
+			:host([hover-effect]) ::slotted(d2l-list-item:hover) {
+				--d2l-list-item-divider-bottom: initial;
+				--d2l-list-item-divider-padding-bottom: initial;
+				--d2l-list-item-divider-padding-top: initial;
+				--d2l-list-item-divider-top: initial;
+			}
+			:host([selectable]) {
```
Comment (dbatiste) [inline components/list/list.js]: I have mixed feelings/thoughts about this.  The list is not selectable, it is the list-items that are selectable.  By having this attribute/property misplaced (imo), I feel we are introducing some awkwardness where the list is trying to override/control rendering inside the items, etc.

File: components/list/list-item.js (hunk 5 above)
Comment (m6jones) [reply components/list/list-item.js]: So why 150 - I was thinking I could make all z-index I use for this component to live in the 100s. I'm happy with best long term solution so changing to meet that I am all for. (Note, I need to double check that all my z-index actually fitting to what I just said)

So to tell you the problem I am solving with the z-index:
So clicking on the list item anywhere should click the checkbox. Unless you are directly on a secondary action. So I needed to put it above the label.

If someone has two actions then the gap between should also click the checkbox.

File: package.json [hunk 10]
```diff
@@ -64,6 +64,7 @@
   "dependencies": {
     "@webcomponents/shadycss": "^1",
     "@webcomponents/webcomponentsjs": "^2",
+    "d2l-inputs": "BrightspaceUI/inputs#semver:^2",
```
Comment (dbatiste) [inline package.json]: Is this used anywhere?  I don't think we should be pulling in.

File: package.json [hunk 11]
```diff
@@ -64,6 +64,7 @@
...
     "d2l-intl": "^2",
     "fastdom": "^1.0.9",
```
Comment (dbatiste) [inline package.json]: Interesting that `fastdom` is here.  I think this branch might be out of date.

File: components/list/list.js (hunk 9 above)
Comment (dlockhart) [reply components/list/list.js]: I agree, plus I really don't like always rendering the checkbox and just hiding it via CSS -- too many extra DOM nodes.

IMO I think we should go with the `selectable` attribute on the items. That way it actually opens the possibility of certain items not being selectable, which I know wasn't a requirement, but would be possible.

File: components/list/list-item.js [hunk 12]
```diff
@@ -169,6 +218,9 @@ class ListItem extends RtlMixin(LitElement) {
 			}
 		});
 	}
+	_handleChange(e) {
+		this.checked = e.target.checked;
```
Comment (dlockhart) [inline components/list/list-item.js]: We probably want to fire an event, something like `d2l-list-item-selection-change`?

File: components/list/list.js (hunk 9 above)
Comment (dlockhart) [reply components/list/list.js]: And in reality, folks will almost always be rendering the items in a `forEach` look anyway, so moving the `selectable` attribute down to the item isn't any more code for them.

Comment (m6jones) [reply components/list/list.js]: So the main reason I put this on the list was to stop the ability have some items being selectable and others not. I haven't talked to Colin yet but having some selectable seems odd to me. I do need to add a disabled attribute if one cannot be selected.

I also, made it so `d2l-list` in PR 4 has the on change event so it is all tied into the parent.

I also, didn't like the extra DOM but when I thought about it it is only adding 2 the `label` and the `input` per item.

File: package.json (hunk 11 above)
Comment (m6jones) [reply package.json]: fastdom was added in the previous PR for the change in breakpoints.

File: components/list/list-item.js (hunk 12 above)
Comment (m6jones) [reply components/list/list-item.js]: See PR 4.

This is only the rendering logic.

File: package.json (hunk 10 above)
Comment (dlockhart) [reply package.json]: Agreed. Once core contains inputs, inputs is going to depend on core so we'd have a circular dependency.

File: package.json (hunk 11 above)
Comment (dlockhart) [reply package.json]: I think we're really hoping to avoid using it in Lit. It shouldn't be necessary.

File: components/list/list.js (hunk 9 above)
Comment (m6jones) [reply components/list/list.js]: So the main reason is to block the ability to make a weird looking list. So if you both think this doesn't matter as much I'll change it :)

The pros of less dom might be better. This change should be pretty easy to do.

File: package.json (hunk 10 above)
Comment (m6jones) [reply package.json]: This is my bad, I didn't remove it when I first tried to just use the checkbox component.

File: package.json (hunk 11 above)
Comment (m6jones) [reply package.json]: You are right! we shouldn't need it. I'll update PR 1 to stop using it. Still getting used to all the new life cycle controls we have but those are what I should be using.

File: components/list/list.js (hunk 9 above)
Comment (dlockhart) [reply components/list/list.js]: Yeah I totally get it. It's just sometimes preventing folks from doing weird things gets outweighed, and this might be one of those cases. Let's see how the conversation about single/multiple/disabled selection of items goes...

File: components/list/list-item.js (hunk 7 above)
Comment (m6jones) [reply components/list/list-item.js]: Oh, miss read this see PR 4 it fixes this issue.

Comment (m6jones) [reply components/list/list-item.js]: Move this discussion to: https://github.com/BrightspaceUI/core/pull/212/files#r324254933

File: package.json (hunk 10 above)
Comment (m6jones) [reply package.json]: - [ ] Remove inputs

File: components/list/list.js (hunk 9 above)
Comment (m6jones) [reply components/list/list.js]: - [ ] Add selectable to list-item

File: components/list/README.md [hunk 13]
```diff
@@ -1,4 +1,4 @@
-<span style="color:red; font-size:24px; text-transform: uppercase;">**In Development**</span>
```
Comment (m6jones) [inline components/list/README.md]: This file has been redone in PR 5/5

File: components/list/test/list.visual-diff.js [hunk 14]
```diff
@@ -17,32 +17,134 @@ describe('d2l-list', function() {
...
+			setTimeout(resolve, 505);
+		});
+	};
+	const tab = (page) => {
+		return new Promise((resolve) => {
+			page.keyboard.press('Tab');
+			setTimeout(resolve, 505);
+		});
+	};
+	const resetMouse = (page) => {
```
Comment (m6jones) [inline components/list/test/list.visual-diff.js]: - [ ] visualDiff has removeFocus that can be used instead.

Comment (m6jones) [reply components/list/test/list.visual-diff.js]: Did remove the resetMouse as it also resets the hover. However, I did replace the resetMouse where it was used to reset the focus.

File: components/list/list.js (hunk 9 above)
Comment (m6jones) [reply components/list/list.js]: I added the selectable to the list-item. Going to mark this as resolved as this seems like the best course of action so far.
